system/jlib/jevent.cpp

```diff
@@ -930,75 +930,621 @@ bool readEvents(const char* filename, IEventVisitor& visitor)
+{
+public:
+    virtual bool visitFile(const char* filename, uint32_t version) override
+    {
+        state = State::Header;
+        openObject(true);
+        openObject(DUMP_STRUCTURE_HEADER, true);
+        doVisitHeader(filename, version);
+        return true;
+    }
+
+    virtual Continuation visitEvent(EventType id) override
+    {
+        if (inHeader())
+        {
+            closeObject();
+            state = State::Events;
+        }
+        if (firstEvent)
+        {
+            openArray(DUMP_STRUCTURE_EVENT, false);
+            openObject(true);
+            firstEvent = false;
+        }
+        else
+            openObject(false);
+        doVisitEvent(id);
         return visitContinue;
     }
+
     virtual bool departEvent() override
     {
-        *out << "departEvent" << std::endl;
+        closeObject();
         return true;
     }
+
     virtual void departFile(uint32_t bytesRead) override
     {
-        *out << "bytesRead: " << bytesRead << std::endl;
+        if (inHeader())
+            closeObject();
+        else if (inEvents())
+            closeArray();
+        openObject(DUMP_STRUCTURE_FOOTER, false);
+        doVisitFooter(bytesRead);
+        closeObject();
+        closeObject(true);
     }
+
 protected:
-    std::ostream* out = &std::cout;
+    virtual void recordAttribute(const char* name, const char* value, bool quoted) override
+    {
+        conditionalDelimiter();
+        indent();
+        appendJSONStringValue(markup, name, value, true, quoted);
+    }
+
 public:
-    CDumpVisitsEventVisitor() {}
-    // An alternate stream may be substituted for testing.
-    CDumpVisitsEventVisitor(std::ostream& _out) : out(&_out) {}
+    using CDumpTextEventVisitor::CDumpTextEventVisitor;
+
+    inline void openObject(bool first)
+    {
+        openObject(nullptr, first);
+    }
+
+protected:
+    void openObject(const char* name, bool first)
+    {
+        openContainer(name, "{ ");
+    }
+
+    void openArray(const char* name, bool first)
+    {
+        openContainer(name, "[ ");
+    }
+
+    void closeArray()
+    {
+        closeContainer("]");
+    }
+
+    void closeObject(bool done = false)
+    {
+        closeContainer("}");
+        dump(done);
+    }
+
+    void openContainer(const char* name, const char* token)
+    {
+        conditionalDelimiter();
+        indent();
+        if (!isEmptyString(name))
+            appendJSONName(markup, name);
+        markup.append(token);
+        firstFlags.push_back(true);
+    }
+
+    void closeContainer(const char* token)
+    {
+        firstFlags.pop_back();
+        indent();
+        markup.append(token);
+    }
+
+    void conditionalDelimiter()
+    {
+        if (!firstFlags.empty())
+        {
+            if (!firstFlags.back())
+                markup.append(",");
+            else
+                firstFlags.back() = false;
+        }
+    }
+
+    void indent()
+    {
+        markup.append('\n');
+        if (!firstFlags.empty())
```

Comment:
minor: Simper as markup.pad(2 * firstFlags.size()); and elsewhere.

Reply:
agree